flask_dance/contrib/gitlab.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import unicode_literals
+
+from flask_dance.consumer import OAuth2ConsumerBlueprint
+from functools import partial
+from flask.globals import LocalProxy, _lookup_app_object
+try:
+    from flask import _app_ctx_stack as stack
+except ImportError:
+    from flask import _request_ctx_stack as stack
+
+
+__maintainer__ = "Justin Georgeson <jgeorgeson@lopht.net>"
+
+
+def make_gitlab_blueprint(
+        client_id=None, client_secret=None, scope=None, redirect_url=None,
+        redirect_to=None, login_url=None, authorized_url=None,
+        session_class=None, backend=None, hostname="gitlab.com"):
+    """
+    Make a blueprint for authenticating with GitLab using OAuth 2. This requires
+    a client ID and client secret from GitLab. You should either pass them to
+    this constructor, or make sure that your Flask application config defines
+    them, using the variables GITLAB_OAUTH_CLIENT_ID and GITLAB_OAUTH_CLIENT_SECRET.
+
+    Args:
+        client_id (str): The client ID for your application on GitLab.
+        client_secret (str): The client secret for your application on GitLab
+        scope (str, optional): comma-separated list of scopes for the OAuth token
+        redirect_url (str): the URL to redirect to after the authentication
+            dance is complete
+        redirect_to (str): if ``redirect_url`` is not defined, the name of the
+            view to redirect to after the authentication dance is complete.
+            The actual URL will be determined by :func:`flask.url_for`
+        login_url (str, optional): the URL path for the ``login`` view.
+            Defaults to ``/gitlab``
+        authorized_url (str, optional): the URL path for the ``authorized`` view.
+            Defaults to ``/gitlab/authorized``.
+        session_class (class, optional): The class to use for creating a
+            Requests session. Defaults to
+            :class:`~flask_dance.consumer.requests.OAuth2Session`.
+        backend: A storage backend class, or an instance of a storage
+                backend class, to use for this blueprint. Defaults to
+                :class:`~flask_dance.consumer.backend.session.SessionBackend`.
+        hostname (str, optional): If using a private instance of GitLab CE/EE,
+            specify the hostname, default is ``gitlab.com``
+
+    :rtype: :class:`~flask_dance.consumer.OAuth2ConsumerBlueprint`
+    :returns: A :ref:`blueprint <flask:blueprints>` to attach to your Flask app.
+    """
+    gitlab_bp = OAuth2ConsumerBlueprint("gitlab", __name__,
+        client_id=client_id,
+        client_secret=client_secret,
+        scope=scope,
+        base_url="https://"+hostname+"/api/v4/",
```

Comment:
I strongly prefer using [string formatting](https://docs.python.org/3/library/string.html#format-string-syntax) rather than concatenation like this. String formatting always coerces variables to strings before putting them together, while concatenation with the plus sign can throw exceptions if the variable is not already a string. You could do something like this:

```python
base_url = "https://{hostname}/api/v4".format(hostname=hostname)
```

However, in this specific case, you have an even better tool available. The [urlobject](https://urlobject.readthedocs.io/en/latest/quickstart.html) library is an elegant way of solving _precisely_ this problem, and Flask-Dance already depends on it. You could do something like this:

```python
base_url = URLObject("https://gitlab.com/api/v4/").with_hostname(hostname)
```

What do you think of these solutions?

Reply:
I prefer using the `format()` function over adding an additional dependency, will make the change.